Support async execution

Grimpper opened the pull request:
Hello!

This PR adds support for asynchronous execution. Whether to load the environments synchronously or asynchronously can be configured with the variable `envrc-async-processing`.

This also refactors the envrc lighter to replace it with a mode line indicator. This indicator introduces a new loading state, where a spinner hints buffers that are waiting for the environment to load. The widget can be disabled by setting the `envrc-add-to-mode-line-misc-info` variable to nil.

The following screencast demonstrates the async processing as well as the new widget:
![envrc-async-processing-demo](https://github.com/user-attachments/assets/7225c5d2-ee52-446b-a50d-cfc94747dc20)

In the future we could consider implementing a [vtable](https://www.gnu.org/software/emacs/manual/html_mono/vtable.html) buffer to display running async processes, similarly to how [dtached.el](https://sr.ht/~niklaseklund/detached.el/) displays detached processes.

Regards,
Sergio.

Grimpper committed:
feat: support asynchronous execution

Grimpper committed:
feat: allow indicator to display loading state while async process is loading environment

Grimpper committed:
refactor: display envrc status in `mode-line-misc-info'

Grimpper committed:
feat: adapt test suite to support asynchronous processing

purcell commented:
Hi, thanks for this. Did you read the discussion in #6 and #53? There are a number of trade-offs around adding this support, and I'm not yet convinced I want to take on async support. As a side comment, changing the lighter code would better be addressed separately: there was discussion of that in #86.

Grimpper commented:
> Hi, thanks for this. Did you read the discussion in #6 and #53? There are a number of trade-offs around adding this support, and I'm not yet convinced I want to take on async support. As a side comment, changing the lighter code would better be addressed separately: there was discussion of that in #86.

I did not know about this discussions. I will take a look. Thanks for pointing them out!

Grimpper commented:
> ... As a side comment, changing the lighter code would better be addressed separately: there was discussion of that in #86.

I decided to move it out from the lighter since that way it will work with mode lines such as doom-modeline out of the box. This is the approach that [notmuch-indicator.el](https://github.com/protesilaos/notmuch-indicator/blob/f399ac107c0d810965ce12fc5499878873bd8414/notmuch-indicator.el#L363) uses.

purcell commented:
> decided to move the status indicator out from the lighter, since that way, it will work with custom mode lines, such as doom-modeline, out of the box. This is the approach that [notmuch-indicator.el](https://github.com/protesilaos/notmuch-indicator/blob/f399ac107c0d810965ce12fc5499878873bd8414/notmuch-indicator.el#L363) uses.

Yeah, not averse to the change itself, just that it's orthogonal to the sync/async change and would be most easily considered/reviewed separately.

Grimpper commented:
> Yeah, not averse to the change itself, just that it's orthogonal to the sync/async change and would be most easily considered/reviewed separately.

Well. I agree that the decision of moving it out of the lighter can be done outside of the async PR, but the status indicator augments the logic to process the loading state, which is an additional feature on top of the async processing. 

If this PR gets merged, I think it's important to provide the loading indicator state. It hints the user which buffers are ready to be used with the environment loaded. 

I think with this indicator we address some of the concerns described in #6, since the user can know what buffers are waiting for their environment to load.

Grimpper committed:
feat: implement denied status

If needed, checking if the envrc directory is allowed, can be implemented with
the following procedures:
```
(defun envrc--allowed-p (env-dir)
  "Return TRUE if ENV-DIR is blocked."
  (let ((allow-hash (envrc--find-allow-hash env-dir))
        (deny-path (concat (or (getenv "XDG_DATA_HOME")
                               (concat (getenv "HOME") "/.local/share"))
                           "/direnv/allow")))
      (locate-file allow-hash
                   (list deny-path))))

(defun envrc--allowed-p (env-dir)
  "Return TRUE if ENV-DIR is blocked."
  (let ((allow-hash (envrc--find-allow-hash env-dir))
        (deny-path (concat (or (getenv "XDG_DATA_HOME")
                               (concat (getenv "HOME") "/.local/share"))
                           "/direnv/allow")))
      (locate-file allow-hash
                   (list deny-path))))
```

Grimpper committed:
feat: when there is a running process prompt the user to kill it

Grimpper commented:
> Yeah, not averse to the change itself, just that it's orthogonal to the sync/async change and would be most easily considered/reviewed separately.

Hello!

I've force pushed some changes to accommodate for the extraction of the status lighter refactor in a separate branch.

There are also some bug fixes to the logic and I've clean up the commits.

In addition I added an option to prompt to query the user before killing already running async processes. I believe this interface solves many concerns about concurrency. For example, if there is a process exporting the environment (eg. `envrc-allow`) and the user issues `envrc-deny`, the process that is running will be killed if the user answers yes to the prompt. Similar logic applies to `envrc-reload` and `envrc-allow`, the user will be given the option to kill the running processes and the new processes will execute. If, for example, the user would have changed the `.envrc` file, it provides the option to load the new environment, without worrying about the callbacks.

cassandracomar commented:
I think, if I'm understanding the discussion on those issues correctly, the open question is what to do about subsequent hooks that depend on the environment loaded by envrc. consider a complex lsp setup where the dependencies don't exist in the process environment until after .envrc is loaded into the process environment (e.g. because it's loaded from a nix flake). running a subsequent lsp-mode hook before the envrc hook actually completes results in a broken state where lsp-mode fails to start correctly. in order to handle this seamlessly, we'd need some way to defer loading subsequent hooks until after envrc finishes loading, and such a mechanism doesn't exist at present.

that said, having this feature as default-off and with this caveat properly documented so users know to reload such hooks after the environment loads is probably sufficient. could we add a post-envrc hook for such users, so they can load their dependent hooks after the async process call completes and the environment is loaded? synchronous loading of .envrc while working on many buffers bites me, personally, all the time -- it would be really nice to be able to work on other buffers while I wait for an envrc environment to load/update.

tylerjl commented:
> [...] having this feature as default-off and with this caveat properly documented so users know to reload such hooks after the environment loads is probably sufficient.

Default-off with caveats listed under the boolean variable seems pretty reasonable? Assuming that the requisites for async execution don't pollute the non-async paths too much and potentially make the default path buggy.

> could we add a post-envrc hook for such users, so they can load their dependent hooks after the async process call completes and the environment is loaded?

Adding a hook that's called once async envrc is complete sounds really nice. For what it's worth, I'm sort of doing this today by adding a variable watch onto `envrc--status` and reacting to when the status changes, and it works as expected (for example, my LSP executables aren't present outside of direnv, but show up after entering it. It's sort of manual since I have to wait for the status to change, check that it's done, and then invoke `eglot`, but that's the tradeoff with the async path).

+1 to separating out the lighter/modeline changes into a separate PR, though.

matthewbauer commented:
Just tested it out a little bit and it looks good to me! I have had my own version of this for a while, but like the idea of it getting merged into mainline envrc. Should be easier to share bug fixes, etc.

purcell commented:
This has now been implemented in #130 and merged: see the new variable `envrc-async`.

It took me a while to figure out a clean design, and the changes involved quite an extensive rewrite. The new implementation is quite different from what was proposed in this PR, and carried over to `ben.el`. In somewhat idiomatic Emacs style, every environment now gets its own separate special buffer in which direnv is run, results are propagated asynchronously via a process sentinel to all corresponding `envrc-mode` buffers, but the default behaviour (`envrc-async = nil`) is still to block.

The unit tests finally run reliably and cover async and sync modes, and I think it *may* even be possible to set `envrc-async` as a directory-local variable just for certain projects, though no promises.

So I'll close this PR now, and thanks all for very slowly nerd-sniping me into this. ❤️

Grimpper commented:
@purcell Hi, congrats on getting this feature implemented.
